### PR TITLE
[RISCV] Generaize reduction tree matching to all integer reductions

### PR DESCRIPTION
**Pull request by preames**
This builds on the transform introduced in https://github.com/llvm/llvm-project/pull/67821, and generalizes it for all integer reduction types.

A couple of notes:
* This will only form smax/smin/umax/umin reductions when zbb is enabled.  Otherwise, we lower the min/max expressions early.  I don't care about this case, and don't plan to address this further.
* This excludes floating point.  Floating point introduces concerns about associativity.  I may or may not do a follow up patch for that case.
* The explodevector test change is mildly undesirable from a clarity perspective.  If anyone sees a good way to rewrite that to stablize the test, please suggest.

### CODE REVIEW

**Review by topperc — APPROVED** [review]
LGTM

**Review by lukel97 — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -11108,6 +11108,31 @@ void RISCVTargetLowering::ReplaceNodeResults(SDNode *N,
...
+  case ISD::SMIN:
+    return ISD::VECREDUCE_SMIN;
+  case ISD::AND:
+    return ISD::VECREDUCE_AND;
+  case ISD::OR:
+    return ISD::VECREDUCE_OR;
+  case ISD::XOR:
+    return ISD::VECREDUCE_XOR;
+  }
+};
```
Comment (qiongsiwu) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:11134]: nit: do we intend to have this extra semicolon after the bracket?

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:11134]: Fixed in 63bbc250440141b1c51593904fba9bdaa6724280

**Comment by asb** [comment]
I reverted this in 824251c9b349d859a9169196cd9533c619a715ce because I was seeing test failures due to the unguarded call to `RISCVSubtarget::getELen` in `combineBinOpOfExtractToReduceTree`, which asserts when V isn't enabled.

I've found you actually can reproduce the crash by just running the fixed-vectors-reduction-formation.ll test without `+v`. `fixed-vectors-reduction-formation.ll` doesn't crash.

Alternatively, here is a bugpoint reduced example derived from 20050604-1.c in the GCC torture suite:
```ll
%union.anon.2.3.19.20 = type { <4 x i16> }

@u = external global %union.anon.2.3.19.20, align 8

declare void @foo() 

define void @main() {
entry:
  %u.promoted.i = load <4 x i16>, ptr @u, align 8
  %add.1.i = add <4 x i16> %u.promoted.i, <i16 24, i16 0, i16 0, i16 0>
  %0 = extractelement <4 x i16> %add.1.i, i64 2
  %tobool = icmp ne i16 %0, 0
  %or.cond = select i1 false, i1 true, i1 %tobool
  %1 = extractelement <4 x i16> %add.1.i, i64 3
  %tobool5 = icmp ne i16 %1, 0
  %or.cond6 = select i1 %or.cond, i1 true, i1 %tobool5
  br i1 %or.cond6, label %if.then, label %if.end

if.then:                                          ; preds = %entry
  unreachable

if.end:                                           ; preds = %entry
  ret void
}

declare void @abort()
```

One obvious fix would be to return immediately from `combineBinOpOfExtractToReduceTree` if `!Subtarget.hasVInstructions()`.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -11126,8 +11151,15 @@ combineBinOpOfExtractToReduceTree(SDNode *N, SelectionDAG &DAG,
...
-  [[maybe_unused]] const unsigned Opc = N->getOpcode();
-  assert(Opc == ISD::ADD && "extend this to other reduction types");
+
+  // TODO: Handle floating point here.
+  if (!VT.isInteger())
+    return SDValue();
+
+  const unsigned Opc = N->getOpcode();
+  const unsigned ReduceOpc = getVecReduceOpcode(Opc);
+  assert(Opc == ISD::getVecReduceBaseOpcode(ReduceOpc) &&
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:11161]: Maybe `isOperationLegalOrCustomOrPromote(ReduceOpc, VT)`?

**Comment by preames** [comment]
Fixed in 199cbec9 and reapplied in 45a334d.

**Pull request merged**